File: hello-swift/Sources/Server/Message.swift
import Foundation

/// JSON message for WebSocket communication.
struct Message: Codable {
    let command: String
    var streamId: String?
    var offset: Int64?
    var length: Int?
    var message: String?
    var status: String?
    var size: Int64?
    var streams: String?

    // MARK: - Factory methods (server → client)

    static func connected(_ connId: String) -> Message {
        Message(command: "CONNECTED", streamId: connId)
    }

    static func created(_ streamId: String) -> Message {
        Message(command: "CREATED", streamId: streamId)
    }

    static func completed(_ streamId: String) -> Message {
        Message(command: "COMPLETED", streamId: streamId)
    }

    static func closed(_ streamId: String) -> Message {
        Message(command: "CLOSED", streamId: streamId)
    }

    static func error(_ msg: String) -> Message {
        Message(command: "ERROR", message: msg)
    }

    static func status(_ streamId: String, _ st: String, _ sz: Int64) -> Message {
        Message(command: "STATUS", streamId: streamId, status: st, size: sz)
    }

    static func streamList(_ ids: [String]) -> Message {
        Message(command: "STREAM_LIST", streams: ids.joined(separator: ","))
    }

    // MARK: - Parse

    /// Parse JSON text into Message + CommandInfo.
    static func parse(_ text: String) throws -> (Message, CommandInfo) {
        guard let data = text.data(using: .utf8) else {
            throw MessageError.invalidJson
        }
        let m = try JSONDecoder().decode(Message.self, from: data)
        guard let info = commandLookup[m.command] else {
            throw MessageError.unknownCommand(m.command)
        }
        return (m, info)
    }

    /// Serialize to JSON string.
    func toJson() -> String {
        let encoder = JSONEncoder()
        guard let data = try? encoder.encode(self),
              let str = String(data: data, encoding: .utf8) else {
            return "{}"
        }
        return str
    }
}

enum MessageError: Error, CustomStringConvertible {
    case invalidJson
    case unknownCommand(String)

    var description: String {
        switch self {
        case .invalidJson: return "Invalid JSON"
        case .unknownCommand(let cmd): return "Unknown command: \(cmd)"
        }
    }
}

File: hello-go/src/server/message.go
package server

import (
	"encoding/json"
	"fmt"
	"strings"
)

// Message is the JSON data transfer object for all control messages.
// Both client→server and server→client use the "command" field.
type Message struct {
	Command  string `json:"command"`
	StreamID string `json:"streamId,omitempty"`
	Offset   *int64 `json:"offset,omitempty"`
	Length   *int   `json:"length,omitempty"`
	Msg      string `json:"message,omitempty"`
	Status   string `json:"status,omitempty"`
	Size     *int64 `json:"size,omitempty"`
	Streams  string `json:"streams,omitempty"`
}

func parseMessage(raw []byte) (*Message, error) {
	var m Message
	if err := json.Unmarshal(raw, &m); err != nil {
		return nil, err
	}
	return &m, nil
}

func (m *Message) toJSON() ([]byte, error) {
	return json.Marshal(m)
}

// Factory helpers — server→client responses.
func msgConnected(connID string) *Message {
	return &Message{Command: "CONNECTED", StreamID: connID, Msg: "Connected"}
}

func msgCreated(id string) *Message {
	return &Message{Command: "CREATED", StreamID: id}
}

func msgCompleted(id string) *Message {
	return &Message{Command: "COMPLETED", StreamID: id}
}

func msgClosed(id string) *Message {
	return &Message{Command: "CLOSED", StreamID: id}
}

func msgStatus(id, status string, size int64) *Message {
	return &Message{Command: "STATUS", StreamID: id, Status: status, Size: &size}
}

func msgStreamList(ids []string) *Message {
	return &Message{Command: "STREAM_LIST", Streams: strings.Join(ids, ",")}
}

func msgError(text string) *Message {
	return &Message{Command: "ERROR", Msg: text}
}

// parseCommand maps the command string to a typed CommandInfo.
func (m *Message) parseCommand() (*CommandInfo, error) {
	switch strings.ToUpper(m.Command) {
	case "CREATE":
		return &CommandInfo{cmdType: CommandTypeStream, streamCmd: CommandCreate}, nil
	case "COMPLETE":
		return &CommandInfo{cmdType: CommandTypeStream, streamCmd: CommandComplete}, nil
	case "CLOSE":
		return &CommandInfo{cmdType: CommandTypeStream, streamCmd: CommandClose}, nil
	case "READ":
		return &CommandInfo{cmdType: CommandTypeData, dataCmd: CommandRead}, nil
	case "GET_STATUS":
		return &CommandInfo{cmdType: CommandTypeQuery, queryCmd: CommandGetStatus}, nil
	case "LIST_STREAMS":
		return &CommandInfo{cmdType: CommandTypeQuery, queryCmd: CommandListStreams}, nil
	default:
		return nil, fmt.Errorf("unknown command: %s", m.Command)
	}
}
